tests/h/views/api_auth_test.py

```diff
@@ -15,6 +17,106 @@
+
+        with pytest.raises(httpexceptions.HTTPBadRequest) as exc:
+            auth_ctrl.get()
+
+        assert 'Unsupported response type' in exc.value.message
+
+    def test_get_redirects_if_user_not_logged_in(self, auth_ctrl,
+                                                 pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy(None)
+        pyramid_request.url = 'http://example.com/auth?client_id=bar'
+
+        with pytest.raises(httpexceptions.HTTPFound) as exc:
+            auth_ctrl.get()
+
+        assert exc.value.location == 'http://example.com/login?next={}'.format(
+                                       url_quote(pyramid_request.url, safe=''))
+
+    def test_get_returns_expected_context(self, auth_ctrl, user_service):
+        ctx = auth_ctrl.get()
+
+        assert ctx == {'username': user_service.fetch.return_value.username,
+                       'client_name': 'Hypothesis',
+                       'client_id': 'valid_id',
+                       'response_type': 'code',
+                       'response_mode': 'web_message',
+                       'state': 'a_random_string'}
+
+    def test_post_returns_auth_code(self, post_auth_ctrl, oauth_service):
+        ctx = post_auth_ctrl.post()
+        assert ctx['code'] == oauth_service.create_grant_token.return_value
+
+    def test_post_sets_origin(self, post_auth_ctrl):
+        ctx = post_auth_ctrl.post()
+        assert ctx['origin'] == 'http://example.com'
+
+    def test_post_returns_state(self, post_auth_ctrl, pyramid_request):
+        ctx = post_auth_ctrl.post()
+        assert ctx['state'] == 'a_random_string'
+
+    @pytest.fixture
+    def auth_ctrl(self, factories, pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy('acct:fred@example.org')
+
+        pyramid_request.GET['client_id'] = 'valid_id'
+        pyramid_request.GET['response_mode'] = 'web_message'
+        pyramid_request.GET['response_type'] = 'code'
+        pyramid_request.GET['state'] = 'a_random_string'
+
+        return views.OAuthAuthorizeController(pyramid_request)
+
+    @pytest.fixture
+    def post_auth_ctrl(self, factories, pyramid_config, pyramid_request):
+        pyramid_config.testing_securitypolicy('acct:fred@example.org')
+
+        pyramid_request.POST['client_id'] = 'valid_id'
+        pyramid_request.POST['response_mode'] = 'web_message'
+        pyramid_request.POST['response_type'] = 'code'
+        pyramid_request.POST['state'] = 'a_random_string'
+
+        return views.OAuthAuthorizeController(pyramid_request)
+
+    @pytest.fixture
+    def oauth_service(self, pyramid_config, pyramid_request):
+        svc = mock.Mock(spec_set=oauth_service_factory(None, pyramid_request))
+        pyramid_config.register_service(svc, name='oauth')
+        return svc
+
+    @pytest.fixture
+    def user_service(self, pyramid_config, pyramid_request):
+        svc = mock.Mock(spec_set=user_service_factory(None, pyramid_request))
```

Comment:
Thanks @seanh - I didn't know about `create_autospec`. The tests worked fine using `create_autospec` but I started a discussion in #backend to clarify the reason for not using it originally.